File: 121.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int maxProfit = 0;
        int len = prices.size();
        vector<int> profit(len, 0);
        int i = 0;
        while (i < len)
        {
            while (i < len - 1 && prices[i] >= prices[i+1])
            {
                ++i;
            }
            int low = prices[i]; //找到一个局部最小值并记录
            while (i < len && prices[i] >= low) //寻找最大利益直到出现比这个局部最小值更小的值即新的局部最小值
            {
                if (prices[i] - low > maxProfit) {
                    maxProfit = prices[i] - low; //更新最大利益
                }
                ++i;
            }
            
        }
        return maxProfit;
    }
};

int main(int argc, char const *argv[])
{
    vector<int> a({7,1,5,3,6,4});
    Solution s = Solution();
    cout << s.maxProfit(a);
    return 0;
}

File: 121.py
class Solution:
    def maxProfit(self, prices: list) -> int:
        if not prices: return 0
        ans = 0
        tmp_min = prices[0]
        # for i in range(len(prices)):
        #     tmp_min = min(tmp_min, prices[i])
        #     ans = max(ans, prices[i] - tmp_min)
        # for p in prices:
        #     tmp_min = min(tmp_min, p)
        #     ans = max(ans, p - tmp_min)
        for p in prices:
            if p < tmp_min:
                tmp_min = p
            else:
                ans = max(ans, p - tmp_min)
        return ans
